core/services/ardupilot_manager/firmware_download/FirmwareDownload.py

```diff
@@ -0,0 +1,206 @@
+
+        Args:
+            url (str): Url to download the file.
+
+        Returns:
+            pathlib.Path: File of the temporary file.
+        """
+
+        filename = FirmwareDownload._generate_random_filename()
+        urlretrieve(url, filename)
+        return filename
+
+    @staticmethod
+    def _validate_firmware(firmware_path: pathlib.Path) -> bool:
+        """A simple validation function for firmware files (apj)
+
+        Args:
+            firmware_path (pathlib.Path): Path of the firmware file;
+
+        Returns:
+            bool: True if valid, False if otherwise.
+        """
+        with open(firmware_path, "r") as firmware_file:
+            data = json.load(firmware_file)
+            keys = data.keys()
+            if "image_size" in keys and "image" in keys:
+                return True
+
+        return False
+
+    def _manifest_is_valid(self) -> bool:
+        """Check if internal content is valid and update it if not.
+
+        Returns:
+            bool: True if valid, False if was unable to validate.
+        """
+        return "format-version" in self._manifest or self.download_manifest()
+
+    def download_manifest(self) -> bool:
+        """Download ArduPilot manifest file
+
+        Returns:
+            bool: True if file was downloaded and validated, False if not.
+        """
+        import gzip
+
+        with urlopen(FirmwareDownload._manifest_remote) as http_response:
+            manifest_gzip = http_response.read()
+            manifest = gzip.decompress(manifest_gzip)
+            self._manifest = json.loads(manifest)
+
+        if "format-version" not in self._manifest:
+            warn("Failed to fetch content of manifest file.", RuntimeWarning)
+            return False
+
+        if self._manifest["format-version"] != "1.0.0":
+            warn("Firmware description file format changed, compatibility may be broken.", RuntimeWarning)
+
+        return True
+
+    def _find_version_item(self, **args: str) -> List[Dict[str, Any]]:
+        """Find version objects in the manifest that match the specific case of **args
+
+        The arguments should follow the same name described in the dictionary inside the manifest
+        for firmware item. Valid arguments can be:
+            mav_type, vehicletype, mav_firmware_version_minor, format, mav_firmware_version_type,
+            platform, latest and others. `-` should be replaced by `_` to use valid python arguments.
+            E.g: `self._find_version_item(vehicletype="Sub", platform="Pixhawk1", mav_firmware_version_type="4.0.1")`
+
+        Returns:
+            List[Dict[str, Any]]: A list of firmware items that match the arguments.
+        """
+        if not self._manifest and not self.download_manifest():
+            return []
+
+        found_version_item = []
+
+        # Make sure that the item matches all args value
+        for item in self._manifest["firmware"]:
+            for key in args.keys():
```

Comment:
```suggestion
            for key in args:
```
FirmwareDownload.py:128:23: C0201: Consider iterating the dictionary directly instead of calling .keys() (consider-iterating-dictionary)

Reply:
Cool